### PR TITLE
feat: implement 1inch swapper for exchange client

### LINKED ISSUE #4836: Implement 1inch for `Swapper2` interface
### Overview

Implement 1inch for `Swapper2` interface

### References and additional details

See `src/lib/swapper/swappers/LifiSwapper/LifiSwapper2.ts` and `src/lib/swapper/swappers/ThorchainSwapper/ThorchainSwapper2.ts`

### Acceptance Criteria

Can perform trades on this swapper exactly as in production.

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

Half day

### PR DESCRIPTION
**Pull request by woodenfurniture**
## Description

Implements 1inch swapper for exchange client. Includes some consolidation and cleanup chores.

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

closes #4836

## Risk

Low risk, but tiny chores affecting evm swappers are included which affects prod potentially.

## Testing

Quick test of EVM swaps to ensure nothing is borked. 

### Engineering

Check 1inch quotes and trades are working as expected. Dont bother with super detailed testing as everything is still in progress and likely to break.

### Operations

See testing notes

## Screenshots (if applicable)

image

### CODE REVIEW

**Comment by woodenfurniture** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #4875** <a href="https://app.graphite.dev/github/pr/shapeshift/web/4875" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/shapeshift/web/4875?utm_source=stack-comment).

File: src/state/apis/swappers/swappersApi.ts [hunk 1]
```diff
@@ -39,7 +40,7 @@ export const swappersApi = createApi({
               swapperName: SwapperName.LIFI,
             })
           } catch (error) {
-            console.error('[getLifiTradeQuoteHelper]', error)
+            console.error(error)
```
Comment (woodenfurniture) [inline src/state/apis/swappers/swappersApi.ts:43]: no need for prefixing console error, now that `logger` is binned we actually get stack traces

File: src/lib/utils/index.ts [hunk 2]
```diff
@@ -16,8 +14,6 @@ import intersection from 'lodash/intersection'
 import isUndefined from 'lodash/isUndefined'
 import union from 'lodash/union'
 
-export * as evm from './evm'
```
Comment (woodenfurniture) [inline src/lib/utils/index.ts:19]: in my opinion this is super annoying and we should refrain from doing it

File: src/lib/swapper/swappers/ThorchainSwapper/ThorchainSwapper.ts [hunk 3]
```diff
@@ -32,7 +32,7 @@ import type {
 } from 'lib/swapper/swappers/ThorchainSwapper/types'
 import { poolAssetIdToAssetId } from 'lib/swapper/swappers/ThorchainSwapper/utils/poolAssetHelpers/poolAssetHelpers'
 import { thorService } from 'lib/swapper/swappers/ThorchainSwapper/utils/thorService'
-import { evm } from 'lib/utils/index'
```
Comment (woodenfurniture) [inline src/lib/swapper/swappers/ThorchainSwapper/ThorchainSwapper.ts:35]: import zoo

File: src/lib/utils/index.ts (hunk 2 above)
Comment (gomesalexandre) [reply src/lib/utils/index.ts:19]: Yeah totally agree with that, finding it hard to navigate the `export * as foo` exports as well

File: src/lib/swapper/swappers/ZrxSwapper/endpoints.ts [hunk 4]
```diff
@@ -81,28 +75,18 @@ export const zrxApi: Swapper2Api = {
...
-    try {
-      const maybeTradeQuoteMetadata = tradeQuoteMetadata.get(tradeId)
-      if (!maybeTradeQuoteMetadata) {
-        return createDefaultStatusResponse(txId)
-      }
-
-      const adapter = assertGetAdapter(maybeTradeQuoteMetadata.chainId)
-      const tx = await adapter.httpProvider.getTransaction({ txid: txId })
-      const status = getTxStatus(tx)
+    // TODO: it might be smart to pass in the chainId rather than having to pull it out of storage
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/ZrxSwapper/endpoints.ts:82]: If the `txId` we're passing was an *actual* txId in the form `accountId*hash*userAddress*memo` (which it isn't currently), then we could use `selectTxById` to select the txId from the store, then deserialize it and get the ChainId from the AccountId.
Interestingly, we don't have a `deserializeTxIndex` function so would have to write one, effectively the opposite of the existing `serializeTxIndex`.

Note however that in its current form, the `txId` we have here is a tx Hash only, so we would have to update `useTradeExecution` to pass down a serialized Txid.

tl;dr passing the `chainId` down would be a much lower lift to start with, and probably more API-y as well

File: src/lib/swapper/swappers/OneInchSwapper/utils/fetchOneInchSwap.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+}: FetchOneInchSwapInput) => {
+  const apiUrl = getConfig().REACT_APP_ONE_INCH_API_URL
+
+  /**
+   * limit of price slippage you are willing to accept in percentage,
+   * may be set with decimals. &slippage=0.5 means 0.5% slippage is acceptable.
+   * Low values increase chances that transaction will fail,
+   * high values increase chances of front running. Set values in the range from 0 to 50
+   */
+  const slippagePercentage = (slippage ? bnOrZero(slippage) : bnOrZero(DEFAULT_SLIPPAGE))
```
Comment (gomesalexandre) [inline src/lib/swapper/swappers/OneInchSwapper/utils/fetchOneInchSwap.ts:37]: In addition to the great comment above, do we want to also rename this to something like `maximumSlippagePercentage` or `requestedSlippagePercentage`?
The reason for this being it easier to grep for *input* / request slippage (i.e what we have currently) and *quote* slippage (which we don't have currently, but may)

**Review by gomesalexandre — APPROVED** [review]
A few comments but conceptually happy with this.

Did a quick sanity check with the flag on:

### It is possible to select the same buy/sell asset

image

Note, this isn't a regression from develop with the flag on and is a holistic issue with the multi-hop swapper

### Minimums aren't handled

Similarly, this was spotted in a previous PR and isn't introduced by this PR, but minimums not being met aren't detected, and clicking "Preview Trade" results in a browser refresh

### Input

image

### Preview

image

### Status

image

### Confirmed state

image

Also regression tested with the flag off:

### Input

image

### Approve

image
image

### Confirm

image

### Status

image

### Confirmed state

image

File: src/lib/swapper/swappers/ZrxSwapper/endpoints.ts (hunk 4 above)
Comment (woodenfurniture) [reply src/lib/swapper/swappers/ZrxSwapper/endpoints.ts:82]: Ah right, in that case it might be wise to rename `txId` to `txHash` to make this clearer. I'll implement that now

File: src/lib/swapper/swappers/OneInchSwapper/utils/fetchOneInchSwap.ts (hunk 5 above)
Comment (woodenfurniture) [reply src/lib/swapper/swappers/OneInchSwapper/utils/fetchOneInchSwap.ts:37]: Also will fix in follow-up

Comment (woodenfurniture) [reply src/lib/swapper/swappers/OneInchSwapper/utils/fetchOneInchSwap.ts:37]: 7dd8621de8c7a1329574050bf20d2e2a702163ec

File: src/lib/swapper/swappers/ZrxSwapper/endpoints.ts (hunk 4 above)
Comment (woodenfurniture) [reply src/lib/swapper/swappers/ZrxSwapper/endpoints.ts:82]: f1351f5feb06d2e1590a7c21dd736ad3e6460179

File: src/lib/utils/index.ts (hunk 2 above)
Comment (0xApotheosis) [reply src/lib/utils/index.ts:19]: I have 0 objection to discontinuing this pattern.

lol at "import zoo".
